src/io/simulator/simulator_handle.hpp

```diff
@@ -0,0 +1,453 @@
+    bool should_drop = drop_threshold < config_.drop_percent;
+
+    if (should_drop) {
+      stats_.dropped_messages++;
+    }
+
+    PromiseKey promise_key{.requester_address = to_address,
+                           .request_id = opaque_message.request_id,
+                           .replier_address = opaque_message.from_address};
+
+    if (promises_.contains(promise_key)) {
+      // complete waiting promise if it's there
+      DeadlineAndOpaquePromise dop = std::move(promises_.at(promise_key));
+      promises_.erase(promise_key);
+
+      bool normal_timeout = config_.perform_timeouts && (dop.deadline < cluster_wide_time_microseconds_);
+
+      if (should_drop || normal_timeout) {
+        stats_.timed_out_requests++;
+        dop.promise.TimeOut();
+      } else {
+        stats_.total_responses++;
+        dop.promise.Fill(std::move(opaque_message));
+      }
+    } else if (should_drop) {
+      // don't add it anywhere, let it drop
+    } else {
+      // add to can_receive_ if not
+      const auto &[om_vec, inserted] = can_receive_.try_emplace(to_address, std::vector<OpaqueMessage>());
+      om_vec->second.emplace_back(std::move(opaque_message));
+    }
+
+    return true;
+  }
+
+  void ShutDown() {
+    std::unique_lock<std::mutex> lock(mu_);
+    should_shut_down_ = true;
+    cv_.notify_all();
+  }
+
+  bool ShouldShutDown() {
+    std::unique_lock<std::mutex> lock(mu_);
+    return should_shut_down_;
+  }
+
+  template <Message Request, Message Response>
+  void SubmitRequest(Address to_address, Address from_address, uint64_t request_id, Request &&request,
+                     uint64_t timeout_microseconds, ResponsePromise<Response> &&promise) {
+    std::unique_lock<std::mutex> lock(mu_);
+
+    uint64_t deadline = cluster_wide_time_microseconds_ + timeout_microseconds;
+
+    std::any message(std::move(request));
+    OpaqueMessage om{.from_address = from_address, .request_id = request_id, .message = std::move(message)};
+    in_flight_.emplace_back(std::make_pair(std::move(to_address), std::move(om)));
+
+    PromiseKey promise_key{.requester_address = from_address, .request_id = request_id, .replier_address = to_address};
+    OpaquePromise opaque_promise(std::move(promise).ToUnique());
+    DeadlineAndOpaquePromise dop{.deadline = deadline, .promise = std::move(opaque_promise)};
+    promises_.emplace(std::move(promise_key), std::move(dop));
+
+    stats_.total_messages++;
+    stats_.total_requests++;
+
+    cv_.notify_all();
+
+    return;
+  }
+
+  template <Message... Ms>
+  requires(sizeof...(Ms) > 0) RequestResult<Ms...> Receive(const Address &receiver, uint64_t timeout_microseconds) {
+    std::unique_lock<std::mutex> lock(mu_);
+
+    uint64_t deadline = cluster_wide_time_microseconds_ + timeout_microseconds;
+
+    while (!should_shut_down_ && (cluster_wide_time_microseconds_ < deadline)) {
+      if (can_receive_.contains(receiver)) {
+        std::vector<OpaqueMessage> &can_rx = can_receive_.at(receiver);
```

Comment:
I wonder whether a `std::queue` wouldn't be a better fit here. Currently this is LIFO, but maybe a FIFO queue should be better.